File: query/src/main/java/com/quarkworks/android/realmtypesafequery/fields/RealmByteArrayField.java
package com.quarkworks.android.realmtypesafequery.fields;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.quarkworks.android.realmtypesafequery.BaseRealmField;
import com.quarkworks.android.realmtypesafequery.interfaces.RealmEmptyableField;

import io.realm.RealmModel;
import io.realm.RealmQuery;

public class RealmByteArrayField<Model extends RealmModel> extends BaseRealmField<Model, byte[]>
        implements RealmEmptyableField<Model,  byte[]> {

    public RealmByteArrayField(@NonNull Class<Model> modelClass, @NonNull String keyPath) {
        super(modelClass, keyPath);
    }

    @Override
    public void equalTo(@NonNull RealmQuery<Model> query, @Nullable byte[] value) {
        if (value == null) {
            this.isNull(query);
            return;
        }

        query.equalTo(this.getKeyPath(), value);
    }

    @Override
    public void notEqualTo(@NonNull RealmQuery<Model> query, @Nullable byte[] value) {
        if (value == null) {
            this.isNotNull(query);
            return;
        }

        query.notEqualTo(this.getKeyPath(), value);
    }

    @Override
    public void never(@NonNull RealmQuery<Model> query) {
        query.equalTo(this.getKeyPath(), new byte[] {}).equalTo(this.getKeyPath(), new byte[] {0});
    }
}

File: query/src/main/java/com/quarkworks/android/realmtypesafequery/fields/RealmByteArrayField.kt
package com.quarkworks.android.realmtypesafequery.fields

import io.realm.RealmModel
import io.realm.RealmQuery

class RealmByteArrayField<Model : RealmModel>(override val modelClass: Class<Model>, override val keyPath: String) :
        RealmField<Model>,
        RealmEquatableField<Model, ByteArray>,
        RealmEmptyableField<Model> {

    override fun equalTo(query: RealmQuery<Model>, value: ByteArray?) {
        if (value == null) {
            this.isNull(query)
            return
        }

        query.equalTo(this.keyPath, value)
    }

    override fun notEqualTo(query: RealmQuery<Model>, value: ByteArray?) {
        if (value == null) {
            this.isNotNull(query)
            return
        }

        query.notEqualTo(this.keyPath, value)
    }

    override fun never(query: RealmQuery<Model>) {
        query.beginGroup().equalTo(this.keyPath, byteArrayOf()).equalTo(this.keyPath, byteArrayOf(0)).endGroup()
    }
}
